tests/suites/test_suite_mpi.function

```diff
@@ -1720,6 +1720,378 @@ exit:
+    TEST_EQUAL( memcmp( da, ra, r.n * sizeof(mbedtls_mpi_uint) ), 0 );
+
+    /* 4a) b += a, cond = 1 => we should get the correct carry */
+    TEST_EQUAL( mbedtls_mpi_core_add_if( da, la, r.n, 1 ), carry );
+
+    /* 4b) and b should have the correct result */
+    TEST_EQUAL( memcmp( da, Xa, r.n * sizeof(mbedtls_mpi_uint) ), 0 );
+
+exit:
+    mbedtls_free( la );
+    mbedtls_free( ra );
+    mbedtls_free( Xa );
+    mbedtls_free( da );
+
+    mbedtls_mpi_free( &X4 );
+    mbedtls_mpi_free( &X8 );
+    mbedtls_mpi_free( &l );
+    mbedtls_mpi_free( &r );
+}
+/* END_CASE */
+
+/* BEGIN_CASE */
+void mbedtls_mpi_core_sub( char * input_l, char * input_r,
+                           char * input_X4, char * input_X8,
+                           int carry )
+{
+    mbedtls_mpi l, r, X4, X8;
+    mbedtls_mpi_uint *la = NULL;
+    mbedtls_mpi_uint *ra = NULL;
+    mbedtls_mpi_uint *Xa = NULL;
+    mbedtls_mpi_uint *da = NULL;
+
+    mbedtls_mpi_init( &l );
+    mbedtls_mpi_init( &r );
+    mbedtls_mpi_init( &X4 );
+    mbedtls_mpi_init( &X8 );
+
+    TEST_EQUAL( mbedtls_test_read_mpi( &l, input_l ), 0 );
+    TEST_EQUAL( mbedtls_test_read_mpi( &r, input_r ), 0 );
+    TEST_EQUAL( mbedtls_test_read_mpi( &X4, input_X4 ), 0 );
+    TEST_EQUAL( mbedtls_test_read_mpi( &X8, input_X8 ), 0 );
+
+    /* All of the inputs are +ve (or zero) */
+    TEST_EQUAL( l.s, 1 );
+    TEST_EQUAL( r.s, 1 );
+    TEST_EQUAL( X4.s, 1 );
+    TEST_EQUAL( X8.s, 1 );
+
+    /* Get the number of limbs we will need */
+    size_t limbs = ( l.n < r.n ) ? r.n : l.n;
+
+    /* We only need to work with X4 or X8, depending on sizeof(mbedtls_mpi_uint) */
+    mbedtls_mpi *X = ( sizeof(mbedtls_mpi_uint) == 4 ) ? &X4 : &X8;
+
+    /* The result shouldn't have more limbs than the longest input */
+    TEST_ASSERT( X->n <= limbs );
+
+    /* Now let's get arrays of mbedtls_mpi_uints, rather than MPI structures */
+    la = mbedtls_calloc( limbs, sizeof(mbedtls_mpi_uint) );
+    ra = mbedtls_calloc( limbs, sizeof(mbedtls_mpi_uint) );
+    Xa = mbedtls_calloc( limbs, sizeof(mbedtls_mpi_uint) );
+    da = mbedtls_calloc( limbs, sizeof(mbedtls_mpi_uint) );
+
+    TEST_ASSERT( la != NULL );
+    TEST_ASSERT( ra != NULL );
+    TEST_ASSERT( Xa != NULL );
+    TEST_ASSERT( da != NULL );
+
+    /* Populate the arrays. As the mbedtls_mpi_uint[]s in mbedtls_mpis (and as
+     * processed by mbedtls_mpi_core_add_if()) are little endian, we can just
+     * copy what we have as long as MSBs are 0 (which they are from calloc())
+     */
+    memcpy( la, l.p, l.n * sizeof(mbedtls_mpi_uint) );
+    memcpy( ra, r.p, r.n * sizeof(mbedtls_mpi_uint) );
+    memcpy( Xa, X->p, X->n * sizeof(mbedtls_mpi_uint) );
+
+    /* 1a) d = l - r => we should get the correct carry */
+    TEST_EQUAL( mbedtls_mpi_core_sub( da, la, ra, limbs ), (mbedtls_mpi_uint) carry );
+
+    /* 1b) d = l - r => we should get the correct result */
+    TEST_EQUAL( memcmp( da, Xa, limbs * sizeof(mbedtls_mpi_uint) ), 0 );
```

Comment:
I'd prefer to use `ASSERT_COMPARE` even when the two sizes are intrinsically the same. One of these days we'll instrument it to show the differences.

`limbs * sizeof(mbedtls_mpi_uint)` could be a variable with a short name (`bytes`, I suppose) to make it less of a chore to type and read.

Reply:
Done.

Any reason why `ASSERT_COMPARE()` doesn't have the same `TEST_` prefix as the other macros?  It's disturbing...

Reply:
Sorry. We started out with `TEST_ASSERT` and I added a bunch of other macros, but I wasn't consistent (you know how me-now and me-three-months-ago can feel like totally different people). I would like to unify the naming, as far as I'm concerned this is technical debt.

Reply:
Once this goes in, if there are no significant other PRs in flight that use the macros without `TEST_`, just write a PR to fix them and I will approve it and find a second reviewer (forcibly, if necessary!). Some tech debt (small/obvious/etc) should be able to be fixed without waiting for "tech debt month". It's good practice to be reworking as we go.

Reply:
Feel free to make that PR. For code in flight, we can keep old names in a transition period. I mainly haven't done this because I'd like someone else to opine on the naming scheme and nobody other than me seemed to care until now.